urlb57eb64056_jacadenac_ISI_Othello_UN_tgz-pJ8-othello_OthelloJ8/src/logica/ia/algoritmosGeneticos/Estrategia.java

```diff
@@ -0,0 +1,435 @@
+		cadena+="\n"+"\t"+ EstrategiaFuncionAptitud.getPesoLibertadesRivalGen(cromosomaMasApto) + "Importancia de dejar sin jugadas al rival";
+		cadena+="\n"+"\t"+ EstrategiaFuncionAptitud.getPesoCasillasCentralesGen(cromosomaMasApto) + "Importancia de obtener las casillas centrales";
+		cadena+="\n"+"Para obtener un peso total de+"+RANGO_GEN; 
+		
+	}
+	
+	// -------------------------------------------------------------------
+	// Este metodo permite guardar en un xml la última población calculada
+	// -------------------------------------------------------------------
+	
+	public static void guardarPoblacion(Genotype Poblacion) throws Exception {
+		DataTreeBuilder builder = DataTreeBuilder.getInstance();
+		IDataCreators doc2 = builder.representGenotypeAsDocument(Poblacion);
+		// Create XML document from generated tree
+		XMLDocumentBuilder docbuilder = new XMLDocumentBuilder();
+		Document xmlDoc = (Document) docbuilder.buildDocument(doc2);
+		XMLManager.writeFile(xmlDoc, new File("PoblacionEstrategia.xml"));
+		
+	}
+	
+	// -------------------------------------------------------------------
+	// Este metodo permite obtiene la población a partir de un archivo
+	// -------------------------------------------------------------------
+	
+	public static Genotype obtenerPoblacion(Configuration conf) throws InvalidConfigurationException{
+		int[][] estrategiasCompetidoras = leerPoblacion();//obtenerCompetidores();
+		Chromosome[] chromosomas= new Chromosome[cantidadCompetidores()];
+		for(int i=0 ; i<chromosomas.length ; i++){
+			Gene[] genes = new Gene[4];
+			for(int j=0 ; j<genes.length ; j++){
+				genes[j] = new IntegerGene(conf, 0, RANGO_GEN);
+				genes[j].setAllele((Integer)(estrategiasCompetidoras[i][j]));
+			}		
+			chromosomas[i] = new Chromosome(conf, genes);	
+			/**/int[] cromosoma = obtenerValoresCromosoma(chromosomas[i]);
+			/**///System.out.println(cromosoma[0]+","+cromosoma[1]+","+cromosoma[2]+","+cromosoma[3]);
+		}
+		Population population = new Population(conf,chromosomas);
+		return new Genotype(conf,population);
+	}
+	
+	public static int cantidadCompetidores(){
+		return leerPoblacion().length;
+	}
+	
+	private static int[] obtenerValoresCromosoma(IChromosome cromosoma) {
+		int[] valores = new int[4];
+		for (int i=0; i<4; i++){
+			valores[i]=getPesoDeGen(cromosoma, i);
+		}
+		return valores;
+	}
+	
+	private static int getPesoDeGen(IChromosome cromosoma, int i) {
+		Integer peso = (Integer) cromosoma.getGene(i).getAllele();
+		return peso;
+	}
+	
+	private static int[][] leerPoblacion() {
+		    int[][] poblacion = new int[17][5];  //tamańo población inicial -> 15
+			JFileChooser fileChooser = new JFileChooser();
+			//String nombreFichero = "poblacionInicial.txt";
+			String nombreFichero = "ultimaGeneracion.txt";
+			File fichero = new File(directorio.getPath(),nombreFichero);
+			fileChooser.setCurrentDirectory(fichero);		
+			try {
+				BufferedReader reader = new BufferedReader(new FileReader(fichero));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *reader*. The resource is referenced at lines 232, 233, 242. The resource is closed at line 245. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *reader* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).